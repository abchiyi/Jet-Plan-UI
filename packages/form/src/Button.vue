<script>
import { h } from "vue";
import { Mask } from "../../mask";
import { ActionFeedback } from "../../action-feedback";
export default {
  name: "m-button",
  data() {
    return {
      value: false,
      maskSize: null
    };
  },
  props: {
    tag: {
      type: String,
      default: "button"
    },
    disabled: {
      type: Boolean,
      default: false
    },
    text: {
      type: Boolean,
      default: false
    },
    hoverAnimation: {
      type: Boolean,
      default: true
    },
    activeAnimation: {
      type: Boolean,
      default: true
    },
    primary: {
      type: Boolean,
      default: false
    },
    radius: {
      type: String,
      default: "s",
      validator: function(value) {
        return ["s", "m", "l"].indexOf(value) !== -1;
      }
    },
    row: {
      type: Boolean,
      default: false
    }
  },
  computed: {
    classes() {
      let classes = ["shape", "m-button"];
      if (this.primary) classes.push("primary");
      if (this.disabled) classes.push("disabled");
      if (this.row) classes.push("row");
      if (this.radius) classes.push("radius-" + this.radius);
      classes.push(this.text ? "text-button" : "button");

      return classes;
    },
    show_mask() {
      return this.hoverAnimation && this.value && !this.disabled;
    }
  },
  methods: {
    enter() {
      this.value = true;
    },
    leave() {
      this.value = false;
    },
    renderDefault() {
      let solt = this.$slots.default;
      if (solt) {
        return solt();
      }
      return "Submit";
    },
    renderMask() {
      return h(Mask, {
        value: this.show_mask
      });
    },
    getSize(el) {
      let bcr = el.getBoundingClientRect();
      return {
        height: bcr.height + "px",
        width: bcr.width + "px"
      };
    }
  },
  render() {
    return h(
      this.activeAnimation ? ActionFeedback : this.tag,
      {
        tag: this.tag,
        class: this.classes,
        onmouseenter: this.enter,
        onmouseleave: this.leave
      },
      {
        default: () => [this.renderMask(), this.renderDefault()]
      }
    );
  }
};
</script>
<style>
/*------------ Default ----------*/

.m-button {
  text-decoration: underline 2px;
  background-color: var(--secondary);
  box-sizing: border-box;
  display: inline-block;
  position: relative;
  padding: 6px 16px;
  -webkit-user-select: none;
  -ms-user-select: none;
  user-select: none;
  font-size: 14px;
  cursor: pointer;
  overflow: hidden;
}

@supports (-webkit-tap-highlight-color: #ffffff00) {
  .m-button {
    -webkit-tap-highlight-color: #ffffff00;
  }
}

.m-button + .m-button {
  margin: 0 0 0 2px;
}

/*------------ Primary ----------*/
.m-button.button.primary {
  background-color: var(--primary);
  color: white;
}

/*------------ Active ----------*/
.m-button.button:active {
  filter: brightness(0.85);
}

/*------------ Animation ----------*/

.m-button {
  transition: all 0.3s var(--ease-out);
}

/*------------ disabled ----------*/

.m-button.disabled:active,
.m-button.disabled {
  color: var(--grey-2);
  cursor: not-allowed;
  filter: unset;
}

/*------------ Disabled Primary  ----------*/
.m-button.button.primary.disabled:active,
.m-button.button.primary.disabled {
  background-color: var(--bgco-1);
  color: var(--grey-0);
}

/*------------ Radius ----------*/

.m-button.radius-s {
  border-radius: 4px;
}
.m-button.radius-m {
  border-radius: 8px;
}

.m-button.radius-l {
  border-radius: 1rem;
}

/*------------ OneLine ----------*/
.m-button.row {
  display: block;
  width: 100%;
}

.m-button.row + .m-button.row {
  margin: 2px 0 0 0;
}

/* Text Button */

/*------------ Default ----------*/

.m-button.text-button {
  background: unset;
  padding: 4px 8px;
}

/*------------ Active ----------*/

.m-button.text-button:active {
  opacity: 0.6;
}

/*------------ Primary ----------*/

.m-button.text-button.primary {
  color: var(--main-0);
}

/* ------------ disabled ---------- */

.m-button.text-button.disabled:active,
.m-button.text-button.disabled {
  color: var(--grey-2);
}

/*------------ Disabled Primary  ----------*/

.m-button.text-button.primary.disabled:active,
.m-button.text-button.primary.disabled {
  color: var(--bgco-1);
  opacity: unset;
}
</style>
