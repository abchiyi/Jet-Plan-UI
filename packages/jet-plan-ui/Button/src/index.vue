<script lang="ts">
import { h, defineComponent, computed } from "vue";
import { ActionFeedback } from "../../ActionFeedback";
// TODO// import { autoTextColor } from "../../tool/lib";
import { JET_THEME } from "../../theme";
export default defineComponent({
  name: "j-button",
  setup() {
    const colors = computed(() => JET_THEME.Theme);

    return {
      colors,
    };
  },
  data() {
    return {
      value: false,
      maskSize: null,
      baseColor: "",
    };
  },
  mounted() {
    function checkAndSetTransition(el: HTMLElement) {
      function getTransitionDuration(el: HTMLElement) {
        return parseFloat(window.getComputedStyle(el).transitionDuration);
      }
      if (!getTransitionDuration(el)) {
        el.style.transition = "0.1s ease-in-out";
      }
    }
    const el = this.$el as HTMLElement;
    el.addEventListener("transitionend", () => {
      checkAndSetTransition(el);
      this.calcTextColor();
    });
    this.calcTextColor();
  },
  updated() {
    this.calcTextColor();
  },
  props: {
    disabled: {
      type: Boolean,
      default: false,
    },
    text: {
      type: Boolean,
      default: false,
    },
    primary: {
      type: Boolean,
      default: false,
    },
    block: {
      type: Boolean,
      default: false,
    },
  },
  methods: {
    calcTextColor() {
      // TODO// import { autoTextColor } from "../../tool/lib";
      // if (!this.text) {
      //   this.baseColor = autoTextColor(this.$refs.self.$el);
      // }
    },
  },
  render() {
    return h(
      ActionFeedback,
      {
        class: [
          "shape",
          "j-button",
          this.text ? "text-button" : "button",
          this.disabled ? "disabled" : null,
          this.primary ? "primary" : null,
          this.block ? "block" : null,
        ],
        tag: "button",
        active: !this.disabled,
        hover: !this.disabled,
        ref: "self",
        style: {
          "--base-text-color": this.baseColor,
        },
      },
      {
        default: () => [this.$slots.default?.()],
      }
    );
  },
});
</script>
<style>
/*------------ Default ----------*/

.j-button {
  border: 0.125em v-bind("colors.border.default") solid;
  background: v-bind("colors.background.dark");
  transition: all 0.3s var(--ease-out);
  color: v-bind("colors.text.default");
  box-sizing: border-box;
  display: inline-block;
  white-space: nowrap;
  position: relative;
  border-radius: 0.25em;
  font-size: 0.8rem;
  padding: 5px 10px;
  user-select: none;
  cursor: pointer;
  margin: 0.25em;
}

/*------------ Focus ----------*/

.j-button:focus-visible {
  outline: 0.2em v-bind("colors.infoColors.info.default") solid;
}

/*------------ disabled ----------*/

.j-button.button.disabled {
  border-color: v-bind("colors.border.light");
  color: v-bind("colors.text.disabled");
  cursor: not-allowed;
}

/*------------ OneLine ----------*/
.j-button.block {
  display: block;
  width: 100%;
}

/* Button Mode Only*/
/*------------ Primary ----------*/
.j-button.primary {
  background-color: v-bind("colors.infoColors.primary.light");
  color: v-bind("colors.text.light");
  border-color: v-bind("colors.infoColors.primary.default");
}

/* Text Button */
/*------------ Default ----------*/
.j-button.text-button {
  background: unset;
  padding: 6px 8px;
  border: unset;
}
/*------------ Primary ----------*/
.j-button.text-button.primary {
  color: v-bind("colors.infoColors.primary.default");
}

/* ------------ disabled ---------- */
.j-button.text-button.disabled {
  color: v-bind("colors.text.disabled");
  background: unset;
}

/* p span */
span > .j-button.text-button,
p > .j-button.text-button {
  padding-top: unset;
  padding-bottom: unset;
}
</style>
