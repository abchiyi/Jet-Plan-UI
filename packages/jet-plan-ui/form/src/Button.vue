<script>
import { h } from "vue";
import { ActionFeedback } from "../../action-feedback";
export default {
  name: "j-button",
  data() {
    return {
      value: false,
      maskSize: null,
    };
  },
  props: {
    disabled: {
      type: Boolean,
      default: false,
    },
    text: {
      type: Boolean,
      default: false,
    },
    primary: {
      type: Boolean,
      default: false,
    },
    row: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    classes() {
      let classes = ["shape", "j-button"];
      if (this.primary) classes.push("primary");
      if (this.disabled) classes.push("disabled");
      if (this.row) classes.push("row");
      classes.push(this.text ? "text-button" : "button");

      return classes;
    },
  },
  methods: {
    renderDefault() {
      let solt = this.$slots.default;
      if (solt) {
        return solt();
      }
      return "Submit";
    },
  },
  render() {
    return h(
      ActionFeedback,
      {
        tag: "button",
        active: true,
        class: this.classes,
      },
      {
        default: () => [this.renderDefault()],
      }
    );
  },
};
</script>
<style>
/*------------ Default ----------*/
.j-button {
  transition: all 0.3s var(--ease-out);
  background-color: var(--secondary);
  border-radius: var(--m-radius);
  box-sizing: border-box;
  display: inline-block;
  position: relative;
  padding: 5px 10px;
  -webkit-user-select: none;
  -ms-user-select: none;
  user-select: none;
  /* color: var(--text-hint); */
  cursor: pointer;
  /* font-size: 1.1rem; */
  text-decoration: underline;
}

.j-button {
  /* margin: 0 0 0 2px; */
  margin: calc(1em * 0.15);
}
/*------------ disabled ----------*/
.j-button.disabled {
  color: var(--text-disabled);
  cursor: not-allowed;
}

/*------------ OneLine ----------*/
.j-button.row {
  display: block;
  width: 100%;
}
.j-button.row + .j-button.row {
  margin: 2px 0 0 0;
}
/* Button Mode Only*/
.j-button.button.primary {
  background-color: var(--primary);
  /* TODO需要在主题文件定义文本色彩 */
  color: white;
}
/* Text Button */
/*------------ Default ----------*/
.j-button.text-button {
  background: unset;
  padding: 6px 8px;
}
/*------------ Primary ----------*/
.j-button.text-button.primary {
  color: var(--primary);
}
/* ------------ disabled ---------- */
.j-button.text-button.disabled {
  color: var(--text-disabled);
}
</style>
