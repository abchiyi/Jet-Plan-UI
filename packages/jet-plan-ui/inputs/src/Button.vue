<script>
import { h } from 'vue';
import { ActionFeedback } from '../../action-feedback';
import { autoTextColor } from '../../tool/lib';
export default {
    name: 'j-button',
    data() {
        return {
            value: false,
            maskSize: null,
            baseColor: '',
        };
    },
    mounted() {
        function getTransitionDuration(el) {
            return parseFloat(window.getComputedStyle(el).transitionDuration);
        }
        function checkAndSetTransition(el) {
            if (!getTransitionDuration(el)) {
                el.style.transition = '0.1s ease-in-out';
            }
        }
        let el = this.$refs.self.$el;
        el.addEventListener('transitionend', () => {
            checkAndSetTransition(el);
            this.calcTextColor();
        });
        this.calcTextColor();
    },
    updated() {
        this.calcTextColor();
    },
    props: {
        disabled: {
            type: Boolean,
            default: false,
        },
        text: {
            type: Boolean,
            default: false,
        },
        primary: {
            type: Boolean,
            default: false,
        },
        row: {
            type: Boolean,
            default: false,
        },
    },
    computed: {
        classes() {
            let classes = ['shape', 'j-button'];
            if (this.primary) classes.push('primary');
            if (this.disabled) classes.push('disabled');
            if (this.row) classes.push('row');
            classes.push(this.text ? 'text-button' : 'button');

            return classes;
        },
    },
    methods: {
        renderDefault() {
            let solt = this.$slots.default;
            if (solt) {
                return solt();
            }
            return 'Submit';
        },
        calcTextColor() {
            if (!this.text) {
                this.baseColor = autoTextColor(this.$refs.self.$el);
            }
        },
    },
    render() {
        return h(
            ActionFeedback,
            {
                class: this.classes,
                tag: 'button',
                active: true,
                ref: 'self',
                style: {
                    '--base-text-color': this.baseColor,
                },
            },
            {
                default: () => [this.renderDefault()],
            }
        );
    },
};
</script>
<style>
/*------------ Default ----------*/

/* a 标签启用下划线 */
a.j-button {
    text-decoration: underline;
}

.j-button {
    transition: all 0.3s var(--ease-out);
    background-color: var(--border-light);
    border-radius: var(--m-radius);
    box-sizing: border-box;
    display: inline-block;
    position: relative;
    padding: 5px 10px;
    -webkit-user-select: none;
    -ms-user-select: none;
    user-select: none;
    color: var(--base-text-color);
    cursor: pointer;
    white-space: nowrap;
    box-shadow: var(--shadow);
}

.j-button {
    margin: calc(1em * 0.15);
}

/* p span */
span > .j-button.text-button,
p > .j-button.text-button {
    padding-top: unset;
    padding-bottom: unset;
}

/*------------ active ----------*/
.j-button.active {
    box-shadow: unset;
}

/*------------ OneLine ----------*/
.j-button.row {
    margin-left: unset;
    margin-right: unset;
    display: block;
    width: 100%;
}

/* Text Button */
/*------------ Default ----------*/
.j-button.text-button {
    background: unset;
    padding: 6px 8px;
}

/*------------ Primary ----------*/
.j-button.primary {
    background-color: var(--primary);
}
.j-button.text-button.primary {
    color: var(--primary);
    background: unset;
}

/*------------ disabled ----------*/
.j-button.disabled {
    background-color: var(--border-light);
    color: var(--text-disabled);
    cursor: not-allowed;
}
.j-button.text-button.disabled {
    color: var(--text-disabled);
    background: unset;
}
</style>
