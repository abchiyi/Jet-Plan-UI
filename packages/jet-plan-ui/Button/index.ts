import type { App, Plugin } from "vue";
import Button from "./src/index.vue";

export const ButtonPlugin: Plugin = {
  install(app: App) {
    app.component("j-button", Button);
  },
};

export { Button };
