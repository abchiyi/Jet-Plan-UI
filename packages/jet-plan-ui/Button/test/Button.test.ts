import { mount } from "@vue/test-utils";
import { expect, describe, test } from "vitest";
import Button from "../src/index.vue";
describe("Button:test", () => {
  test("SnapShots", () => {
    const wrapper = mount(Button, {
      slots: {
        default: () => "slot default",
      },
    });

    expect(wrapper.html()).matchSnapshot();
  });

  test("Slot:default", () => {
    const wrapper = mount(Button, {
      slots: {
        default: "default slot",
      },
    });

    expect(wrapper.text()).toBe("default slot");
  });

  test("Class", async () => {
    const wrapper = mount(Button, {
      slots: {
        default: () => "slot default",
      },
    });

    expect(wrapper.classes()).include("shape");
    expect(wrapper.classes()).include("j-button");
    expect(wrapper.classes()).include("button");

    expect(wrapper.classes()).not.include("disabled");
    expect(wrapper.classes()).not.include("block");

    await wrapper.setProps({
      disabled: true,
      block: true,
    });

    expect(wrapper.classes()).include("disabled");
    expect(wrapper.classes()).includes("block");
  });

  const StyleTypes = ["primary", "warning", "danger", "success"];
  test.each(StyleTypes)("Props:styleType", async styleType => {
    const wrapper = mount(Button);
    expect(wrapper.classes()).not.include(styleType);
    await wrapper.setProps({ styleType });
    expect(wrapper.classes()).include(styleType);
  });

  const Mode = ["text", "button", "outline"];
  test.each(Mode)("Props:mode", (mode: any) => {
    const wrapper = mount(Button, { props: { mode } });
    expect(wrapper.classes()).include(mode);
  });
});
