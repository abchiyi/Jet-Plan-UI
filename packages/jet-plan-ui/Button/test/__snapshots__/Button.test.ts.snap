// Vitest Snapshot v1

exports[`Button:test > SnapShots 1`] = `
"<button data-v-03123537=\\"\\" class=\\"j-base-action j-action-feedback hover shape j-button button\\" ontouchstart=\\"function () { [native code] }\\" ontouchcancel=\\"function () { [native code] }\\" ontouchend=\\"function () { [native code] }\\">slot default<transition-stub name=\\"mask\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
    <!---->
  </transition-stub></button>"
`;
